clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -129,17 +126,15 @@ else if (this.subscriptionType != type)
     }
 
     public boolean subscribe(Set<String> topics, ConsumerRebalanceListener listener) {
-        if (listener == null)
-            throw new IllegalArgumentException("RebalanceListener cannot be null");
-
-        setSubscriptionType(SubscriptionType.AUTO_TOPICS);
-
-        this.rebalanceListener = listener;
+        registerRebalanceListener(listener);
```

Comment:
Why put this outside the lock?

Reply:
The `ConsumerRebalanceListener` is given by the user, so it can be anything, therefore I didn't want to lock it. Do you think we should lock it?

Reply:
Hmm, but we're not actually calling the listener here. We do that separately.

Reply:
Ok, that's a fair point. I'll make it snychronized.